### DISCUSSION TITLE
Adding "Company Administrator" and "Everyone except external users"?

### DISCUSSION THREAD

**Discussion by plamber** [pnp/pnpcore#1121] [Q&A] [answered]
Hello,
Could you hint on how we can add the "Company Administrators" and "Everyone except external users" claims to a SharePoint site?

Reference documentation: https://pnp.github.io/pnpcore/using-the-sdk/security-users.html?q=everyone%20except%20external%20users

Both have a unique ID at the end of the name that I still need to figure out how to get with the SDK.

Thank you for your feedback,
Patrick

**Comment by jansenbe** [comment, marked as answer]
@plamber : you can do an `EnsureUser` using those names:

```csharp
var everyoneExceptEnternalUsers = await context.Web.EnsureUserAsync("Everyone except external users");
var companyAdministators = await context.Web.EnsureUserAsync("Company Administrator");
```


Also see https://pnp.github.io/pnpcore/using-the-sdk/security-users.html#addingensuring-a-user for more context.

**Reply by plamber** [reply]
HI @jansenbe,
Thank you for your response but there is a minor challenge with multi-language use cases. Let us imagine you have a German site, the "Everyone except eternal users" does not work since the display name is translated. To be 100% sure to get it right for all languages, we should find a way to use the loginname with the GUID generated for each tenant.

c:0-.f|rolemanager|spo-grid-all-users/[guid]

The Company Administrator seems to be the same in all languages.

Cheers

**Reply by jansenbe** [reply]
@plamber : you've a point there, which reminded me of this which I did for page transformation a long time ago: https://github.com/pnp/pnpframework/blob/dev/src/lib/PnP.Framework/Modernization/Extensions/WebExtensions.cs#L342-L509. Such a similar approach could be taken here as well. Can you create an issue with the relevant details so we can track it? If you want to create a PR that would speed up things :)

**Reply by plamber** [reply]
Hi,
I created an issue to track it. https://github.com/pnp/pnpcore/issues/1125

I would implement the same thing for the CompanyAdministrator. Do you know the GUID at the end of the admin claim?

I am wondering. Could we expose the AuthenticationRealm when initializing an SPWeb without performing an additional call? Are we unable to retrieve the realm name from there when initializing SPWeb? https://github.com/pnp/pnpframework/blob/a113663f37eb30cfea3cbe6047b73770961daa76/src/lib/PnP.Framework/Extensions/SecurityExtensions.cs#L1207-L1215



Cheers

**Reply by jansenbe** [reply]
`PnPContext.GlobalOptions.AADTenantId` contains what you need and will be prepopulated in telemetry is not disabled via `PnPContext.SetAADTenantId` during context initialization (`PnPContextFactory.ConfigureTelemetry`)

**Reply by jansenbe** [reply]
Feature was added, see the EnsureEveryoneExceptExternalUsers methods on IWeb
